src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -194,15 +187,32 @@ public virtual void GetPolicies(JsonIgnoreCondition? ignoreCondition)
 
         public bool IsPropertyPolicy { get; protected set; }
 
-        // The name from a Json value. This is cached for performance on first deserialize.
-        public byte[]? JsonPropertyName { get; set; }
-
-        // The name of the property with any casing policy or the name specified from JsonPropertyNameAttribute.
-        public byte[]? Name { get; private set; }
+        // There are 3 copies of the property name:
+        // 1) NameAsString. The unescaped property name.
+        // 2) NameAsUtf8. The Utf8 version of NameAsString. Used during during deserialization for property lookup.
+        // 3) EscapedName. The escaped verson of NameAsString and NameAsUtf8 written during serialization. Internally shares
+        // the same instances of NameAsString and NameAsUtf8 if there is no escaping.
+
+        /// <summary>
+        /// The unescaped name of the property.
+        /// Is either the actual CLR property name,
+        /// the value specified in JsonPropertyNameAttribute,
+        /// or the value returned from PropertyNamingPolicy(clrPropertyName).
+        /// </summary>
         public string? NameAsString { get; private set; }
 
-        // Key for fast property name lookup.
-        public ulong PropertyNameKey { get; set; }
+        /// <summary>
+        /// Utf8 version of NameAsString.
+        /// </summary>
+        public byte[]? NameAsUtf8 { get; private set; }
+
+        /// <summary>
+        /// The escaped name passed to the writer.
+        /// </summary>
+        /// <remarks>
+        /// JsonEncodedText is a value type so a field is used (not a property) to avoid unnecessary copies.
+        /// </remarks>
+        public JsonEncodedText? EscapedName;
```

Comment:
Should this be a property?

Reply:
If it's a property then every call to the getter will create a new `JsonEncodedText`. This was actually an issue that was fixed in 3.1 for performance.